kubernetes/src/main/java/io/kubernetes/client/Watch.java

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package io.kubernetes.client;
+
+import com.google.gson.annotations.SerializedName;
+import com.squareup.okhttp.ResponseBody;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Iterator;
+
+/**
+ * Watch class implements watch mechansim of kubernetes. For every list API
+ * call with a watch parameter you should be able to pass its call to this class
+ * and watch changes to that list. For example CoreV1Api.listNamespace has watch
+ * parameter, so you can create a call using CoreV1Api.listNamespaceCall and
+ * set watch to True and watch the changes to namespaces.
+ */
+public class Watch<T> implements Iterable<Watch.Response<T>> {
+
+    public static class Response<T> {
+        @SerializedName("type")
+        public String type;
+
+        @SerializedName("object")
+        public T object;
+
+        Response(String type, T object) {
+            this.type = type;
+            this.object = object;
+        }
+    }
+
+    Type watchType;
+    ResponseBody response;
+    JSON json;
+
+    public Watch(JSON json, ResponseBody body, Type watchType) throws IOException, ApiException {
+        this.response = body;
+        this.watchType = watchType;
+        this.json = json;
+    }
+
+    public Response<T> next() throws IOException {
+        String line = response.source().readUtf8Line();
+        if (line == null) {
+            return null;
+        }
+        return json.deserialize(line, watchType);
+    }
+
+    public Iterator<Response<T>> iterator() {
+        return new Iterator<Response<T>>() {
+            Response<T> nextItem = null;
+            boolean loadNext = true;
+            IOException lastException = null;
+            public boolean hasNext() {
+                if (loadNext) {
+                    try {
+                        nextItem = Watch.this.next();
```

Comment:
I am calling `hasNext()` in `next()` method so there is no need for user to call this before `next()`. I faced two options here, to raise exception on `next()` or `hasNext()`. I chose this way so when I use for-each with this iterator and the watch closes because of timeout, it won't raise an exception (because for-each calls hasNext() before next()). Of course if user cares about what happen they can always call next() and get the reason (or just call next() on Watch class itself).